File: Proyecto Mis Finanzas Python/src/gestorAplicación/interno/ahorros.py
from .cuenta import Cuenta
from excepciones import accountsException
from datetime import datetime
import random

class Ahorros(Cuenta):
    #Atributos de clase
    _cuentasAhorroTotales = []

    #Constructor
    def __init__(self, **kwargs):
        #Atributos de instancia
        Ahorros._cuentasAhorroTotales.append(self)
        if "saldo" in kwargs:
            self._saldo = kwargs.pop("saldo")
        super().__init__(**kwargs)

#REVISAR SOBRECARGA

    def crearCuenta(self, banco, clave, nombre, **kwargs):
        if "divisa" in kwargs:
            return Ahorros(banco = banco, clave = clave, nombre = nombre, divisa = kwargs["divisa"])
        else:
            return Ahorros(banco = banco, clave = clave, nombre = nombre)
    
    def __str__(self):
        return "Cuenta: " + self._nombre + "\nCuentas de Ahorros # " + str(self._id) + "\nTitular: " + self.getTitular().getNombre() + "\nBanco: " + self._banco.getNombre() + "\nDivisa: " + str(self._divisa.value) + "\nSaldo: " + str(self._saldo) + " " + str(self._divisa.value)
    
    def __contains__(self, elementos):
        for elemento in elementos:
            if self.getId() == elemento.getId():
                return elementos.index(elemento)
        else:
            return None
        
    # Método de la funcionalidad asesoramiento de inversiones
    def vaciar_cuenta(self, gota):
        from .movimientos import Movimientos
        from .categoria import Categoria
        movimiento = Movimientos(origen = self, destino = gota, cantidad = self.getSaldo(), categoria = Categoria.OTROS, fecha = datetime.now())
        Movimientos.getMovimientosTotales().remove(movimiento)
        self.setSaldo(0)

    # Funcionalidad de Suscripciones de Usuarios
    def invertirSaldo(self):
        from .movimientos import Movimientos
        from .categoria import Categoria
        probabilidad = self.getTitular().getSuscripcion().getProbabilidadInversion()
        rand = random.random()+probabilidad
        print()
        if(rand >= 1 and self.getSaldo() != 0):
            return(Movimientos.crearMovimiento(self, self.getSaldo() + self.getSaldo() * probabilidad, Categoria.FINANZAS, datetime.now()))
        else:
            raise accountsException.FailedInvestmentException(self.getTitular())


    # Funcionalidad Prestamos
    @classmethod
    def comprobarPrestamo(cls,_cuentas):
        cuentasPrestamos = []
        bancos = []

        for cuenta in _cuentas:
            prestamos = cuenta.getBanco().getPrestamo()
            if(prestamos>0):
                cuentasPrestamos.append(cuenta)
            else:
                bancos.append(cuenta.getBanco())
        if len(cuentasPrestamos)!=0:
            return cuentasPrestamos
        else:
            return bancos

    @staticmethod
    def limpiarPropiedades(cls, arreglo):
        arreglo.remove("cuentasTotales")
        #Verificar que otras variables se crean
    #Verificar su uso

    #Realizar el método del compare to
    def compareTo(self, cuenta):
        if self.getSaldo() > cuenta.getSaldo():
            return 1
        elif self.getSaldo() < cuenta.getSaldo():
            return -1
        else:
            return 0

    #Métodos Get & Set
    def getSaldo(self):
        return self._saldo
    def setSaldo(self, saldo):
        self._saldo = saldo

    @classmethod
    def getCuentasAhorrosTotales(cls):
        return cls._cuentasAhorroTotales
    @classmethod
    def setCuentasAhorrosTotales(cls, cuentasAhorrosTotales):
        cls._cuentasAhorroTotales = cuentasAhorrosTotales

File: Proyecto Mis Finanzas Java/src/gestorAplicación/interno/Ahorros.java
package gestorAplicación.interno;

import gestorAplicación.externo.Banco;
import gestorAplicación.externo.Divisas;
import java.util.ArrayList;
import java.util.Date;
import java.lang.reflect.Field;
import java.time.Instant;

public class Ahorros extends Cuenta{
	//Atributos
	private static final long serialVersionUID = 7L;
	public static final String nombreD = "Ahorros";
	private static ArrayList<Ahorros> cuentasAhorroTotales = new ArrayList<Ahorros>();;
	private double saldo = 0.0d;

	//Constructores
	public Ahorros(Banco banco, int clave, Divisas divisa, String nombre, Double saldo) {
		super(banco, clave, divisa, nombre);
		this.saldo = saldo;
		Ahorros.getCuentasAhorroTotales().add(this);
	}
	
	public Ahorros(Banco banco, int clave, String nombre, Double saldo) {
		super(banco, clave, nombre);
		this.saldo = saldo;
		Ahorros.getCuentasAhorroTotales().add(this);
	}
	
	public Ahorros(Double saldo) {
		super();
		this.saldo = saldo;
		Ahorros.getCuentasAhorroTotales().add(this);
	}
	
	public Ahorros(Banco banco, int clave, Divisas divisa, String nombre) {
		super(banco, clave, divisa, nombre);
		Ahorros.getCuentasAhorroTotales().add(this);
	}	
	
	public Ahorros(Banco banco, int clave, String nombre) {
		super(banco, clave, nombre);
		Ahorros.getCuentasAhorroTotales().add(this);
	}
	
	public Ahorros() {
		super();
		Ahorros.getCuentasAhorroTotales().add(this);
	}
	
	//Métodos
	public Ahorros crearCuenta(Banco banco, int clave, Divisas divisa, String nombre) {
		return (new Ahorros(banco, clave, divisa, nombre));
	}
	
	public Ahorros crearCuenta(Banco banco, int clave, String nombre) {
		return (new Ahorros(banco, clave, nombre));
	}
	
	//Funcionalidad de Suscripciones de Usuarios
	public Object invertirSaldo() {
		float probabilidad = this.getTitular().getSuscripcion().getProbabilidad_Inversion();
		double rand = (double)((Math.random()) + probabilidad);
		if(rand >= 1){
			Movimientos m = (Movimientos) Movimientos.crearMovimiento(this, this.getSaldo() + this.getSaldo() * probabilidad, Categoria.FINANZAS, new Date());
			return (m);
		}else {
			return("Su inversion ha fallado, inténtelo de nuevo. Considere subir de nivel para aumentar la probabilidad de tener inversiones exitosas");
		}
	}
	
	// Funcionalidad asesor inversiones
	public void vaciarCuenta(Ahorros gota) {
		Movimientos movimiento = new Movimientos(this, gota, this.getSaldo(), Categoria.OTROS,
				Date.from(Instant.now()));
		this.getTitular().getMovimientosAsociados().add(movimiento);
		Movimientos.getMovimientosTotales().remove(movimiento);
	}
	
	public static void limpiarPropiedades(ArrayList<String> arreglo) {
		arreglo.remove("serialVersionUID");
		arreglo.remove("nombreD");
		arreglo.remove("cuentasAhorroTotales");
	}
	
	public int CompareTo(Ahorros cuenta) {
		if(this.getSaldo() > cuenta.getSaldo()) {
			return 1;
		}
		else if(this.getSaldo() < cuenta.getSaldo()) {
			return -1;
		}
		else {
			return 0;
		}
	}
	
	public String toString() {
		return "Cuenta: " + super.nombre +
				"\nCuenta de Ahorros # " + this.id +
				"\nBanco: " + this.banco.getNombre() +
				"\nTitular: " + this.getTitular().getNombre() +
				"\nDivisa: " + this.divisa +
				"\nSaldo: " + this.saldo + " " + this.divisa;
	}
	
	public double getSaldo() {
		return saldo;
	}
	
	public void setSaldo(Double saldo) {
		this.saldo = saldo;
	}
	
	public static ArrayList<Ahorros> getCuentasAhorroTotales() {
		return cuentasAhorroTotales;
	}

	public static void setCuentasAhorroTotales(ArrayList<Ahorros> cuentasAhorroTotales) {
		Ahorros.cuentasAhorroTotales = cuentasAhorroTotales;
	}

	//	Funcionalidad Prestamo
	public static ArrayList<?> comprobarPrestamo(ArrayList<?> cuentas){
		ArrayList<Ahorros> cuentasPrestamo = new ArrayList<Ahorros>();
		ArrayList<String> bancos = new ArrayList<String>();

		//Pasamos por todas las cuentas del usuario y comprobamos que el prestamo sea diferente de 0
		for(int i=0;i<cuentas.size();i++){
			Double prestamo = ((Cuenta) cuentas.get(i)).getBanco().getPrestamo();
			if(prestamo>0){
				cuentasPrestamo.add((Ahorros) cuentas.get(i));
			}else{
				bancos.add(((Cuenta) cuentas.get(i)).getBanco().getNombre());
			}
		}

		if(cuentasPrestamo.size()!=0){
			return cuentasPrestamo;
		}else{
			return bancos;
		}
	}



}
